File: sdk-rust/crates/dex-sdk/tests/integ/basic_immutable_step_options_workflow.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use dex_sdk::{
    Context, Flow, HandlerError, HandlerResult, RetryPolicy, Step, StepDecision, StepList,
    StepMovement, StepOptions, Wait, WaitForFailurePolicy,
};

pub(crate) struct BasicImmutableStepOptionsWorkflow {
    start: StartStep,
    failing_wait: FailingWaitStep,
}

impl BasicImmutableStepOptionsWorkflow {
    pub(crate) fn new() -> Self {
        Self {
            start: StartStep,
            failing_wait: FailingWaitStep,
        }
    }
}

impl Flow for BasicImmutableStepOptionsWorkflow {
    type StartInput = i32;

    fn steps(&self) -> StepList<'_, Self::StartInput> {
        StepList::start(&self.start).and(&self.failing_wait)
    }
}

struct StartStep;

impl Step for StartStep {
    type Input = i32;

    fn execute(&self, _context: &mut Context, _input: i32) -> HandlerResult<StepDecision> {
        let options = StepOptions::new()
            .wait_for_retry(RetryPolicy::new().maximum_attempts(1))
            .wait_for_failure(WaitForFailurePolicy::Proceed);
        Ok(StepDecision::go_to_many([StepMovement::to_with_options(
            &FailingWaitStep,
            1,
            options,
        )]))
    }
}

struct FailingWaitStep;

impl Step for FailingWaitStep {
    type Input = i32;

    fn wait_for(&self, _context: &mut Context, input: i32) -> HandlerResult<Wait> {
        Err(HandlerError::new(
            "BasicImmutableStepOptionsFailure",
            format!("expected wait failure {input}"),
        ))
    }

    fn execute(&self, context: &mut Context, input: i32) -> HandlerResult<StepDecision> {
        if !context.wait_for_method_failed() {
            return Err(HandlerError::new(
                "BasicImmutableStepOptionsFailure",
                "wait failure was not reported",
            ));
        }
        if input == 1 {
            return Ok(StepDecision::go_to(self, 2));
        }
        Ok(StepDecision::graceful_complete(input))
    }

    fn options(&self) -> StepOptions<Self::Input> {
        StepOptions::new()
            .wait_for_retry(RetryPolicy::new().maximum_attempts(1))
            .wait_for_failure(WaitForFailurePolicy::FailFlow)
    }
}

File: sdk-java/src/test/java/io/superdurable/dex/integ/BasicImmutableStepOptionsWorkflow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.RetryPolicy;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepMovement;
import io.superdurable.dex.StepOptions;
import io.superdurable.dex.Wait;
import io.superdurable.dex.WaitForFailurePolicy;

final class BasicImmutableStepOptionsWorkflow implements Flow<Integer> {
    private final StartStep start = new StartStep();
    private final FailingWaitStep failingWait = new FailingWaitStep();

    @Override
    public StepList<Integer> getSteps() {
        return StepList.startStep(start).otherSteps(failingWait);
    }

    final class StartStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            final StepOptions override = StepOptions.newBuilder()
                    .waitForRetry(RetryPolicy.newBuilder().maximumAttempts(1).build())
                    .waitForFailure(WaitForFailurePolicy.PROCEED)
                    .build();
            return StepDecision.goToMulti(
                    StepMovement.of(FailingWaitStep.class, 1, override));
        }
    }

    final class FailingWaitStep implements Step<Integer> {
        @Override
        public Class<Integer> getInputType() {
            return Integer.class;
        }

        @Override
        public Wait waitFor(final Context context, final Integer input) {
            throw new IllegalStateException("expected wait failure " + input);
        }

        @Override
        public StepDecision execute(final Context context, final Integer input) {
            if (!context.waitForMethodFailed()) {
                throw new IllegalStateException("wait failure was not reported");
            }
            if (input == 1) {
                return StepDecision.goTo(FailingWaitStep.class, 2);
            }
            return StepDecision.gracefulComplete(input);
        }

        @Override
        public StepOptions getStepOptions() {
            return StepOptions.newBuilder()
                    .waitForRetry(RetryPolicy.newBuilder().maximumAttempts(1).build())
                    .waitForFailure(WaitForFailurePolicy.FAIL_FLOW)
                    .build();
        }
    }
}
